core/node/via_btc_watch/src/message_processors/votable.rs

```diff
@@ -0,0 +1,131 @@
+use sqlx::types::chrono::{DateTime, Utc};
+use via_btc_client::{
+    indexer::BitcoinInscriptionIndexer,
+    types::{BitcoinTxid, FullInscriptionMessage},
+};
+use zksync_dal::{Connection, Core, CoreDal};
+use zksync_types::{aggregated_operations::AggregatedActionType, L1BatchNumber, H256};
+
+use super::{MessageProcessor, MessageProcessorError};
+
+const DEFAULT_THRESHOLD: f64 = 0.5;
+
+#[derive(Debug)]
+pub struct VotableMessageProcessor {
+    threshold: f64,
+}
+
+impl VotableMessageProcessor {
+    pub fn new() -> Self {
+        Self {
+            threshold: DEFAULT_THRESHOLD,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl MessageProcessor for VotableMessageProcessor {
+    async fn process_messages(
+        &mut self,
+        storage: &mut Connection<'_, Core>,
+        msgs: Vec<FullInscriptionMessage>,
+        indexer: &mut BitcoinInscriptionIndexer,
+    ) -> Result<(), MessageProcessorError> {
+        // Get the current timestamp
+        let dt = Utc::now();
+        let naive_utc = dt.naive_utc();
+        let offset = dt.offset().clone();
+        let dt = DateTime::<Utc>::from_naive_utc_and_offset(naive_utc, offset);
+
+        for msg in msgs {
+            if let Some(l1_batch_number) = indexer.get_l1_batch_number(&msg).await {
+                match msg {
+                    FullInscriptionMessage::ProofDAReference(proof_msg) => {
+                        let tx_id = convert_txid_to_h256(proof_msg.common.tx_id);
+                        let mut votes_dal = storage.via_votes_dal();
+
+                        votes_dal
```

Comment:
I would definitely go with this and explicitly check if `l1_batch_number.0 == last_inserted_block + 1`, otherwise with the current logic we could miss voting for some batch.

Reply:
I created this logic in the last commit, please check it out

Reply:
I did, but I see that you added this check `if l1_batch_number.0 <= last_inserted_block` which still doesn't guarantee that some blocks won't be skipped, right?

Reply:
fixed